src/jitlayers.cpp

```diff
@@ -607,12 +555,235 @@ CompilerResultT JuliaOJIT::CompilerT::operator()(Module &M)
     return CompilerResultT(std::move(ObjBuffer));
 }
 
+void jl_register_jit_object(const object::ObjectFile &debugObj,
+                            std::function<uint64_t(const StringRef &)> getLoadAddress,
+                            std::function<void *(void *)> lookupWriteAddress);
+
+#ifdef JL_USE_JITLINK
+
+namespace {
+
+using namespace llvm::orc;
+
+struct JITObjectInfo {
+    std::unique_ptr<MemoryBuffer> BackingBuffer;
+    std::unique_ptr<object::ObjectFile> Object;
+    StringMap<uint64_t> SectionLoadAddresses;
+};
+
+class JLDebuginfoPlugin : public ObjectLinkingLayer::Plugin {
+    std::map<MaterializationResponsibility *, std::unique_ptr<JITObjectInfo>> PendingObjs;
+    // Resources from distinct MaterializationResponsibilitys can get merged
+    // after emission, so we can have multiple debug objects per resource key.
+    std::map<ResourceKey, std::vector<std::unique_ptr<JITObjectInfo>>> RegisteredObjs;
+
+public:
+    void notifyMaterializing(MaterializationResponsibility &MR, jitlink::LinkGraph &G,
+                             jitlink::JITLinkContext &Ctx,
+                             MemoryBufferRef InputObject) override
+    {
+        // Keeping around a full copy of the input object file (and re-parsing it) is
```

Comment:
How bad is the performance/memory impact?

Reply:
I haven't measured it, but it's the same as before in the RTDyld implementation – apparently even for ELF, where `getObjectForDebug` isn't a no-op, a full copy is made internally.

Reply:
Shouldn't we still pretend that `getObjectForDebug` is meaningfully implemented however?

Reply:
Where do you mean? It's RTDyld only, and we do use it there if available.

Reply:
Looking at the upcoming plugin support in llvm/lib/ExecutionEngine/Orc/DebuggerSupportPlugin.cpp, it appears good support for this will appear in the future, so perhaps not worth working on now.

Alternatively, could we enable split-debug, and parse the .dwo files instead, instead of making JITLink deal with implementing the extract-dwo objcopy pass itself?

Reply:
We can't use DebuggerSupportPlugin verbatim. I actually planned on using that first – hence the whole LLVM 14 effort –, but it already patches the section addresses to match the runtime memory layout, so we can't re-parse it using `llvm::ObjectFile`. In the long run, particularly as more platforms will move onto JITLink, it will certainly be a good idea to remove the `ObjectFile` references from `debuginfo` and keep track of the required tables/… at a finer granularity.

Reply:
As for split DWARF information, I haven't really thought about how that would interact with OrcJIT at all.